### PR TITLE
Add stepper interface

### LINKED ISSUE #10903: Stepper control
NuttX lacks an interface to handle stepper motor (motor interface is too heavy and complicated for such motor). Controlling the stepper from the userspace is also costly/complicated depending on the stepper driver behind (i.e. A4988 requires multiples GPIOs to configure it).

### PR DESCRIPTION
**Pull request by leducp**
## Summary
Add a stepper generic upper-half driver
Add a stepper A4988 lower-half driver

close #10903

### CODE REVIEW

**Comment by leducp** [comment]
Note: the code in rp2040_bringup.c is here to show how to add a stepper in a board. It shall be removed **before** merge hence the reason why this PR is in draft.

I plan to add a stepper example application too when the stepper interface will be merged.

**Comment by xiaoxiang781216** [comment]
should we put it into https://github.com/apache/nuttx/tree/master/drivers/motor? Please fix ci issue too, @leducp .

**Comment by leducp** [comment]
> should we put it into https://github.com/apache/nuttx/tree/master/drivers/motor? Please fix ci issue too, @leducp .

Gladly! I wasn't sure of the approach since the stepper interface is different from the current motor.h one (which is more BLDC/BDC centric).

File: Kconfig [hunk 1]
```diff
@@ -2034,6 +2034,39 @@ config DEBUG_MOTOR_INFO
 
 endif # DEBUG_MOTOR
 
+config DEBUG_STEPPER
+	bool "Stepper Debug Features"
```
Comment (acassis) [inline Kconfig:2038]:
```suggestion
	bool "Stepper Motor Debug Features"

File: Kconfig [hunk 2]
```diff
@@ -2034,6 +2034,39 @@ config DEBUG_MOTOR_INFO
...
+	default n
+	depends on STEPPER
+	---help---
+		Enable motor debug features.
+
+if DEBUG_STEPPER
+
+config DEBUG_STEPPER_ERROR
+	bool "Stepper Error Output"
```
Comment (acassis) [inline Kconfig:2047]:
```suggestion
	bool "Stepper Motor Error Output"

File: Kconfig [hunk 3]
```diff
@@ -2034,6 +2034,39 @@ config DEBUG_MOTOR_INFO
...
+	default n
+	depends on DEBUG_ERROR
+	---help---
+		Enable motor error output to SYSLOG.
+
+config DEBUG_STEPPER_WARN
+	bool "Stepper Warnings Output"
```
Comment (acassis) [inline Kconfig:2054]:
```suggestion
	bool "Stepper Motor Warnings Output"

File: Kconfig [hunk 4]
```diff
@@ -2034,6 +2034,39 @@ config DEBUG_MOTOR_INFO
...
+	default n
+	depends on DEBUG_WARN
+	---help---
+		Enable motor warning output to SYSLOG.
+
+config DEBUG_STEPPER_INFO
+	bool "Stepper Informational Output"
```
Comment (acassis) [inline Kconfig:2061]:
```suggestion
	bool "Stepper Motor Informational Output"

**Review by acassis — APPROVED** [review]

**Comment by acassis** [comment]
@leducp really cool idea!

Please include later an apps/example/stepper and some basic documentation

**Comment by xiaoxiang781216** [comment]
> > should we put it into https://github.com/apache/nuttx/tree/master/drivers/motor? Please fix ci issue too, @leducp .
> 
> Gladly! I wasn't sure of the approach since the stepper interface is different from the current motor.h one (which is more BLDC/BDC centric).

@raiden00pl do you have any suggestion?

**Review by jerpelea — COMMENTED** [review]
please add a commit message

**Comment by leducp** [comment]
> please add a commit message

Note sure what you means: my commits have messages. Do you want a particular message? I Could add it when I'll squash to go out of the draft state.

**Comment by leducp** [comment]
> @leducp really cool idea!
> 
> Please include later an apps/example/stepper and some basic documentation

Do you have any advice to show how to integrate it in a board? For now I have a raw example in the rp2040_bringup.c file, but this is temporary and shall not be merged. I could add snippet in a documentation file maybe? Or in the example folder?

**Comment by raiden00pl** [comment]
The idea behind `motor.h` was to create a universal interface for various types of motors so I think it's better to modify it to fit better for stepper motors. You can just add another configuration option for motor interface:

```
config MOTOR_UPPER_HAVE_STEPS
	bool "Have steps control"
	default n
```

then extend `struct motor_params_s` with:

```
#ifdef CONFIG_MOTOR_UPPER_HAVE_STEPS
  uint32_t steps; 
  int16_t steps_speed;
  uint16_t microstep;
#endif
```

Currently `motor.h` uses floats which is not ideal for chips without FPU and I would like to add support for fixedmath someday. But if you need only step control, you can just set `CONFIG_MOTOR_UPPER_HAVE_STEPS` and all float values should be not included. It's possible to control stepper motor also with the FOC algorithm, in this case the concept of 'steps' makes no sense.

Using `motor.h`, the stepper motor driver can later implement control methods other than step if needed and having the same interface, we can more easily change the type of motor used (better modularity).

**Comment by leducp** [comment]
> The idea behind `motor.h` was to create a universal interface for various types of motors so I think it's better to modify it to fit better for stepper motors. You can just add another configuration option for motor interface:
> 
> ```
> config MOTOR_UPPER_HAVE_STEPS
> 	bool "Have steps control"
> 	default n
> ```
> 
> then extend `struct motor_params_s` with:
> 
> ```
> #ifdef CONFIG_MOTOR_UPPER_HAVE_STEPS
>   uint32_t steps; 
>   int16_t steps_speed;
>   uint16_t microstep;
> #endif
> ```
> 
> Currently `motor.h` uses floats which is not ideal for chips without FPU and I would like to add support for fixedmath someday. But if you need only step control, you can just set `CONFIG_MOTOR_UPPER_HAVE_STEPS` and all float values should be not included. It's possible to control stepper motor also with the FOC algorithm, in this case the concept of 'steps' makes no sense.
> 
> Using `motor.h`, the stepper motor driver can later implement control methods other than step if needed and having the same interface, we can more easily change the type of motor used (better modularity).

I understand the idea but I feel that it became a kind of god object anti pattern...
The whole motor interface works with a different mindset like start and stop. Moreover, what if one need to control a stepper and a bldc? Both interface will be mixed in the code no?

About your last point, if one want to control a stepper as a motor, it have to implement a specific control in the motor interface and not in the stepper one (to match this mindset).

**Comment by raiden00pl** [comment]
> I understand the idea but I feel that it became a kind of god object anti pattern...
The whole motor interface works with a different mindset like start and stop. Moreover, what if one need to control a stepper and a bldc? Both interface will be mixed in the code no?

If you want control stepper motor with PWM peripheral, for example to use `CONFIG_PWM_PULSECOUNT` functionality then you need start and stop. In your case you are directly controlling the GPIO, but the same can be achieved with PWM if your chip can precisely generate pulses or you just want to control the motor shaft speed.

When you have different types of motors to control, then the motor 'lower-half' logic can just ignore parameters that are not needed (e.g. stepper motor in step mode doesn't care about 'float position;'). This probably requires some modifications to the "upper-half" motor driver, but the interface itself will not be a problem here.

You say it's mixing interfaces and anti pattern, I say it's creating a generic motor solution. Depends on how you look at the problem :)

> About your last point, if one want to control a stepper as a motor, it have to implement a specific control in the motor interface and not in the stepper one (to match this mindset). I'm pretty sure that it is NOT possible to implement FOC control for stepper with a stepper driver like the A4988.

Correct, FOC is definitely not possible with A4988, my comment was more about the fact that a stepper motor doesn't necessarily have to be controlled in 'steps' unit. Someone may even prefer control stepper with `float position;` where the motor shaft movement is in radian.

**Comment by leducp** [comment]
> > I understand the idea but I feel that it became a kind of god object anti pattern...
> > The whole motor interface works with a different mindset like start and stop. Moreover, what if one need to control a stepper and a bldc? Both interface will be mixed in the code no?
> 
> If you want control stepper motor with PWM peripheral, for example to use `CONFIG_PWM_PULSECOUNT` functionality then you need start and stop. In your case you are directly controlling the GPIO, but the same can be achieved with PWM if your chip can precisely generate pulses or you just want to control the motor shaft speed.
> 
I don't understand why you'll need start/stop with pulse count? Is this not precisely an option that let you ask a job with X pulses and the system will stop on is own?  

> When you have different types of motors to control, then the motor 'lower-half' logic can just ignore parameters that are not needed (e.g. stepper motor in step mode doesn't care about 'float position;'). This probably requires some modifications to the "upper-half" motor driver, but the interface itself will not be a problem here.
> 
> You say it's mixing interfaces and anti pattern, I say it's creating a generic motor solution. Depends on how you look at the problem :)
> 
Yes I understand this point, but my point is: stepper is its own kind of beast and the whole motor interface is cumbersome/heavy for the basic usage of a stepper : moving X steps in one direction. 
What is the point to have a generic interface to match all possible aspects if the user application is heavier and need to be adapted when changing the control mode? (by adding steps in integer in this interface).

**Comment by raiden00pl** [comment]
> I don't understand why you'll need start/stop with pulse count? Is this not precisely an option that let you ask a job with X pulses and the system will stop on is own?

Never mind. You are right, PWM is stopped automatically by the 'lower-half' PWM pulse count driver at least in the STM32 implementation. 

> Yes I understand this point, but my point is: stepper is its own kind of beast and the whole motor interface is cumbersome/heavy for the basic usage of a stepper : moving X steps in one direction.
What is the point to have a generic interface to match all possible aspects if the user application is heavier and need to be adapted when changing the control mode? (by adding steps in integer in this interface).

The point is that a stepper motor is a motor like any other. You control torque/force, speed or position. In stepper case position is additionally quantized, but it's the same kind of device.
The generic interface helps with modularity and allows you to write generic applications. An extreme example of the opposite approach can be seen in `drivers/sensors`, where without a standarized interface we ended up with a complete mess.

The functionality of the driver that you won't use will be optimized by the compiler. If something is unnecessary for a given application (e.g. you don't need `MTRIOC_SET_LIMITS`), you can disable it with Kconfig options.

The stepper driver you are using is a simple STEP/DIR device, but there are drivers on the market with additional features accessible over SPI (L6470, L6482) or I2C (PCA9629A). With them you can control position, speed, speed ramp or get current position. Which driver should support these chips ? `stepper.c` or `motor.c` ? If `stepper.c` then more functionality from `motor.c` will need to be ported to `stepper.c`.

**Comment by leducp** [comment]
> > I don't understand why you'll need start/stop with pulse count? Is this not precisely an option that let you ask a job with X pulses and the system will stop on is own?
> 
> Never mind. You are right, PWM is stopped automatically by the 'lower-half' PWM pulse count driver at least in the STM32 implementation.
> 
> > Yes I understand this point, but my point is: stepper is its own kind of beast and the whole motor interface is cumbersome/heavy for the basic usage of a stepper : moving X steps in one direction.
> > What is the point to have a generic interface to match all possible aspects if the user application is heavier and need to be adapted when changing the control mode? (by adding steps in integer in this interface).
> 
> The point is that a stepper motor is a motor like any other. You control torque/force, speed or position. In stepper case position is additionally quantized, but it's the same kind of device. The generic interface helps with modularity and allows you to write generic applications. An extreme example of the opposite approach can be seen in `drivers/sensors`, where without a standarized interface we ended up with a complete mess.
> 
I would like to play the devil's advocate on this point: yes this is a mess but there is full support for a bunch a real hardware, unlike the current motor interface. Probably not a fair point but at this moment NuttX motor.h have no use without custom patches to add real hardware on it and the current interface didn't help to add simple devices that are widespread (almost all 3D printers have basic stepper driver).

> The functionality of the driver that you won't use will be optimized by the compiler. If something is unnecessary for a given application (e.g. you don't need `MTRIOC_SET_LIMITS`), you can disable it with Kconfig options.
> 
Yes and no: if you need both in one application, you'll end-up with it in both interface.

> The stepper driver you are using is a simple STEP/DIR device, but there are drivers on the market with additional features accessible over SPI (L6470, L6482) or I2C (PCA9629A). With them you can control position, speed, speed ramp or get current position. Which driver should support these chips ? `stepper.c` or `motor.c` ? If `stepper.c` then more functionality from `motor.c` will need to be ported to `stepper.c`.
Yes, but I think that is the main point where I disagree: even for this more complex stepper drivers, floating math or fixed point math are too complicated, and the only big difference is the possibility to let the motor spin indefinitely... no torque control or fancy classic motor things here.
So in this specific case, current motor interface is still too complex imho (I really think many applications that use steppers just want a KISS interface, with easy to integrate lower half).

Another point that I saw is that theses complexes drivers ends with already good low level userspace interface (SPI, I2C, PWM for ESC) and many peoples just put the control code in their application (maybe not the best move but still well supported) and that is not the case for GPIO that are really chip specific (and the user space GPIO interface is really bothersome to use).

**Comment by acassis** [comment]
> > @leducp really cool idea!
> > Please include later an apps/example/stepper and some basic documentation
> 
> Do you have any advice to show how to integrate it in a board? For now I have a raw example in the rp2040_bringup.c file, but this is temporary and shall not be merged. I could add snippet in a documentation file maybe? Or in the example folder?

Just some basic application example to control the direction, then number of steps to each direction and also changing the micro-stepping values. Just print some text before executing each action to let the user know what is supposed to happen.

**Comment by leducp** [comment]
> > > @leducp really cool idea!
> > > Please include later an apps/example/stepper and some basic documentation
> > 
> > 
> > Do you have any advice to show how to integrate it in a board? For now I have a raw example in the rp2040_bringup.c file, but this is temporary and shall not be merged. I could add snippet in a documentation file maybe? Or in the example folder?
> 
> Just some basic application example to control the direction, then number of steps to each direction and also changing the micro-stepping values. Just print some text before executing each action to let the user know what is supposed to happen.

```c
int main(int argc, FAR char *argv[])
{
  int rc;

  if (argc < 2)
  {
    printf("usage: stepper count (positive: CW, negative: CCW)");
    return 1;
  }

  int count = atoi(argv[1]);

  int fd = open("/dev/stepper0", O_RDWR);
  if (fd < 0)
  {
    printf("Oops: %s\n", strerror(errno));
    exit(1);
  }

  printf("GO -> %d\n", count);

  rc = ioctl(fd, STEPIOC_MICROSTEPPING, 8);
  if (rc < 0)
  {
    printf("STEPIOC_MICROSTEPPING: %s\n", strerror(errno));
    exit(2);
  }

  struct stepper_state_s state;
  struct stepper_job_s job;

  job.steps = count;
  job.speed = 8000; /* step / s */
  rc = write(fd, &job, sizeof(struct stepper_job_s)); /* blocking */
  if (rc != sizeof(struct stepper_job_s))
  {
    printf("write: %s\n", strerror(errno));
  }

  rc = read(fd, &state, sizeof(struct stepper_state_s));
  if (rc != sizeof(struct stepper_state_s))
  {
    printf("read: %s\n", strerror(errno));
  }

  printf("Current position: %ld\n", state.position);

  return 0;
}
```

It may need a bit more refine but theses should do the trick I think.

**Comment by acassis** [comment]
@leducp @raiden00pl I think there are pros and cons of both options:

1) A generic driver is good to have a standard interface that could share at least some basic features;
2) A specific driver is easy to follow and understand, because we don't need to care about code from other drivers (avoiding my #ifdefs)

I suggest we could start with the current approach for this driver model before forcing it to adapt to a generic concept. I think the stepper driver itself will evolve a little bit to support other solutions. For example, a user could just use 4 GPIOs connected to a H-Bridge and a timer to control a stepper motor directly or some other device as @raiden00pl cited.

This way the drive will become more generic over the time and eventually it will be easier to integrate both. I know the immediate thought is the driver could become more complex over the time, but I think to support more models it will become more generic.

**Comment by raiden00pl** [comment]
> This way the drive will become more generic over the time and eventually it will be easier to integrate both. I know the immediate thought is the driver could become more complex over the time, but I think to support more models it will become more generic.

That's what I mean. If `stepper.c` is developed further we will end up with something very similar to `motor.c`.

File: drivers/motor/Kconfig [hunk 5]
```diff
@@ -62,3 +62,23 @@ endif
...
 endif # MOTOR
+
+menuconfig STEPPER
+	bool "Stepper control drivers"
+	default n
+
+if STEPPER
+
+config STEPPER_UPPER
+	bool "Motor generic upper-half driver"
```
Comment (acassis) [inline drivers/motor/Kconfig:73]:
```suggestion
	bool "Stepper Motor Generic upper-half driver"

File: drivers/motor/Kconfig [hunk 6]
```diff
@@ -62,3 +62,23 @@ endif
...
+	default n
+	---help---
+		Enables building of a stepper generic upper half driver.
+
+config STEPPER_A4988
+	bool "A4988 stepper driver"
```
Comment (acassis) [inline drivers/motor/Kconfig:79]:
```suggestion
	bool "A4988 Stepper Motor Driver"

**Comment by acassis** [comment]
> > This way the drive will become more generic over the time and eventually it will be easier to integrate both. I know the immediate thought is the driver could become more complex over the time, but I think to support more models it will become more generic.
> 
> That's what I mean. If `stepper.c` is developed further we will end up with something very similar to `motor.c`.

Yes, so let bring his contribution and evolve it incrementally

**Pull request merged**